Code/Framework/AzFramework/AzFramework/Input/Mappings/InputMapping.cpp

```diff
@@ -9,9 +9,156 @@
+            }
+        }
+        return AZStd::vector<AZStd::string>(uniqueInputDeviceTypes.begin(), uniqueInputDeviceTypes.end());
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////
+    const AZStd::vector<AZStd::string> InputMapping::InputChannelNameFilteredByDeviceType::GetValidInputChannelNamesBySelectedDevice() const
+    {
+        AZStd::vector<AZStd::string> validInputChannelNames;
+        InputDeviceId selectedDeviceId(m_inputDeviceType.c_str());
+        InputDeviceRequests::InputChannelIdSet validInputChannelIds;
+        InputDeviceRequestBus::Event(selectedDeviceId,
+                                     &InputDeviceRequests::GetInputChannelIds,
+                                     validInputChannelIds);
+        for (const InputChannelId& inputChannelId : validInputChannelIds)
+        {
+            validInputChannelNames.push_back(inputChannelId.GetName());
+        }
+        AZStd::sort(validInputChannelNames.begin(), validInputChannelNames.end());
+        return validInputChannelNames;
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////
+    void InputMapping::ConfigBase::Reflect(AZ::ReflectContext* context)
+    {
+        if (AZ::SerializeContext* serializeContext = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serializeContext->Class<InputMapping::ConfigBase>()
+                ->Version(0)
+                ->Field("Output Input Channel Name", &InputMapping::ConfigBase::m_outputInputChannelName)
+            ;
+
+            if (AZ::EditContext* editContext = serializeContext->GetEditContext())
+            {
+                editContext->Class<InputMapping::ConfigBase>("Input Mapping: Base",
+                    "Maps multiple different input sources to a single output input channel.")
+                    ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
+                        ->Attribute(AZ::Edit::Attributes::AutoExpand, true)
+                    ->DataElement(AZ::Edit::UIHandlers::Default, &ConfigBase::m_outputInputChannelName, "Output Input Channel Name",
+                        "The unique input channel name (ie. input event name) output by the input mapping.\n"
+                        "This will be truncated if its length exceeds that of InputChannelId::MAX_NAME_LENGTH = 64")
+                        ->Attribute(AZ::Edit::Attributes::Max, InputChannelId::MAX_NAME_LENGTH)
+                ;
+            }
+        }
+
+        InputChannelNameFilteredByDeviceType::Reflect(context);
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////
+    AZStd::shared_ptr<InputMapping> InputMapping::ConfigBase::CreateInputMappingAndAddToContext(InputContext& inputContext) const
+    {
+        AZStd::shared_ptr<InputMapping> inputMapping = CreateInputMapping(inputContext);
+        if (inputMapping)
```

Comment:
question: Shouldn't this always succeed? Wondering why the check is needed, or can this function also return nullptr potentially

Reply:
This can potentially return nullptr, see for example InputMappingAnd::Config::CreateInputMapping and InputMappingOr::Config::CreateInputMapping